Ad4630 remove avg 0

machschmitt opened the pull request:
## PR Description

ADI engineers have reported a hardware bug when using average mode zero (no averaging) with AD4630 devices and requested to disallow using that average mode. 
This PR contains a commit to remove average mode zero from the list of available average modes and make average mode one (average over 2 samples) the default average mode on device startup.
This PR also includes a fix from @nunojsa which testers said to be needed as well.

## PR Type
- [x] Bug fix (a change that fixes an issue)
- [ ] New feature (a change that adds new functionality)
- [ ] Breaking change (a change that affects other repos or cause CIs to fail)

## PR Checklist
- [x] I have conducted a self-review of my own code changes
- [x] I have tested the changes on the relevant hardware (@ladace reported it has been tested and deemed good)
- [x] I have updated the documentation outside this repo accordingly (if there is the case) (no need to update docs)

nunojsa commented:
This looks good to me. Only thing is my fix that I would prefer not to be backported into main. Reason is that fix does not apply in the upstream version of the spi-engine driver. So, instead of backporting the fix, I would prefer you to migrate that project into using [spi-axi-spi-engine-ex.c](https://github.com/analogdevicesinc/linux/blob/main/drivers/spi/spi-axi-spi-engine-ex.c). It's more work but it should not be that bad. Main thing is to properly update the spi transfers .len and .bits_per_word to match the spi subsystem expectations. In the legacy driver that was a mess...

From a quick look, it seems you also need the new spi_optimize() thing @dlech upstreamed. I guess he can also help you if needed.

Note the goal is to migrate all the projects using spi-engine to the new drivers so that we can drop the legacy one...

machschmitt commented:
Updated to new spi message optimize mechanism and to use new spi-engine driver.
Please, hold this PR for a little bit. @ladace is experimenting with some additional changes to this.
 
> Note the goal is to migrate all the projects using spi-engine to the new drivers so that we can drop the legacy one...

humm, changing to the new spi-engine looks pretty straightforward (at least for ad4630). We may be able to update them all with coccinelle or even some `sed` or `awk`. Having it all tested would be a harder task then I think ...

dlech commented on drivers/iio/adc/ad4630.c:
```diff
@@ -720,8 +720,8 @@ static int ad4630_buffer_preenable(struct iio_dev *indio_dev)
 	spi_bus_lock(st->spi->master);
-	spi_engine_offload_load_msg(st->spi, &st->offload_msg);
-	spi_engine_offload_enable(st->spi, true);
+	spi_optimize_message(st->spi, &st->offload_msg);
```
This needs a matching call to `spi_unoptimize_message()` somewhere to avoid potential resource leaks.

dlech replied:
Also, since this message doesn't change, `spi_optimize_message()` can be moved to right after `spi_message_init_with_transfers(&st->offload_msg, &st->offload_xfer, 1);` so that it doesn't have to get called multiple times.

dlech replied:
Also, also, we still need `spi_engine_ex_offload_load_msg()` here.

dlech replied:
> In fact, it can change. ... (unless the spi controller does caching etc...)

Ah yes, you are right. And `spi_optimize_message()` does cause the axi-spi-engine driver to compile the commands and store them, so we do need to be careful to call it in the right place.

dlech commented on drivers/iio/adc/ad4630.c:
```diff
@@ -1480,6 +1480,11 @@ static int ad4630_probe(struct spi_device *spi)
 		ad4630_set_pga_gain(indio_dev, 0);
 	}
 
+	/* Set default averaging mode to 2 samples */
```
Would be helpful to future readers to mention the reason for this.

dlech reviewed (commented):
Line 11174 (`st->offload_xfer.len = 1;`) need to be moved to right before or after `st->offload_xfer.bits_per_word = st->bits_per_word;` and changed to `st->offload_xfer.len = (st->bits_per_word > 16) ? 4 : 2;`.

The "old" spi engine driver has a bug where it treats `len` as words instead of bytes so when we move to the "new" spi engine driver, we need to account for this fix.

Probably won't notice the bug if only using the message with offload but still would be good to fix it for correctness.

machschmitt commented:
Don't like to have these forever testing PRs but please don't merge this yet.
@ladace is helping with testing this. ADAQ4224 seems to work properly with the current version.
I made the changes suggested.
A new commit to update dts to board revision 3.
One additional new experimental commit on top.
According to @ladace, this has to be tested with additional max31827 changes.
I'm seeing if I can get those and the ones here into a single rebased branch.

Change log v2 -> v3:
- Updated `offload_xfer.len` according to review comments
- Registered devm `spi_unoptimize_message()` action for `spi_optimize_message()`
- Added comment about average mode 2 as default
- Updated device trees to use, upstream like, spi-engine driver implementation
- Added commit to limit lane rate to 1.7 MSPS 
- Updated dts to board revision 3
  + vref from 5V to 4.096V
  + Updated i2c controller register address and interrupt number
  + More properties to max31827
  + New eeprom device node
  + AD4224 lane-mode to 0 (interleaved) and clock-mode to 0 (spi-compatible)
  + Removed AD4224 adi,dual-data-rate property

dlech commented:
> * Registered devm `spi_unoptimize_message()` action for `spi_optimize_message()`

This doesn't sound right. Since `spi_optimize_message()` is called in `ad4630_buffer_preenable()`, then to be balanced, `spi_unoptimize_message()` needs to be called in `ad4630_buffer_postdisable()`.

And as Nuno pointed out in correcting me, `spi_unoptimize_message()` should not be called `ad4630_prepare_spi_sampling_msg()`.

nunojsa commented:
> This doesn't sound right. Since spi_optimize_message() is called in ad4630_buffer_preenable(), then to be balanced, spi_unoptimize_message() needs to be called in ad4630_buffer_postdisable().

Definitely not... I actually don't think we would end up with some umbalanced optimization but still looks like unneeded calls. In fact, since the right thing  (I believe)  is to optimize on buffer enable and unoptimize on disable, should we actually care about optimizations in here? Seems to me that we're not gaining much...

dlech commented on drivers/iio/adc/ad4630.c:
```diff
@@ -1159,19 +1163,24 @@ static int ad4630_pwm_get(struct ad4630_state *st)
 	 * to mark the transfer as a read.
 	 */
 	st->offload_xfer.speed_hz = AD4630_SPI_SAMPLING_SPEED;
 	st->offload_xfer.rx_buf = (void *)-1;
```
yes, this is still needed.

nunojsa replied:
Got it... It would be nice to somehow make it more explicit. Anyways, nothing to care here

dlech commented:
> should we actually care about optimizations in here? Seems to me that we're not gaining much...

It's not about optimization in this case but the fact that this function compiles the message into SPI Engine instructions. It's a bit hacky to have to call it manually but it saves us from having to make `spi_maybe_[un]optimize_message()` in spi.c public to be able to call it automatically in the axi-spi-engine driver if it wasn't called already.

We could refactor it though if it is throwing people off and causing problems.

machschmitt commented on drivers/iio/adc/ad4630.c:
```diff
@@ -1287,7 +1287,7 @@ static int ad4630_config(struct ad4630_state *st)
 		return ret;
 
-	if (lane_mode == AD4630_ONE_LANE_PER_CH && data_rate &&
+	if (lane_mode == AD4630_ONE_LANE_PER_CH && !data_rate &&
```
Yeah, indeed. Should have waited until had been properly tested. Removed this patch on v4.

ladace replied:
This is a valid change, If there is no DDR, max sampling frequency should be limited to 1.7MSPS. This applies for 30bits averaging mode and 24bits differential + 8 bits common modes on ad4630 1 lane per channel

nunojsa replied:
Not saying is not valid. Just saying it was unrelated with the commit it was on. It needs to be on a separate commit...

machschmitt commented on drivers/iio/adc/ad4630.c:
```diff
@@ -1197,7 +1206,14 @@ static void ad4630_prepare_spi_sampling_msg(struct ad4630_state *st,
 	}
 
 	st->offload_xfer.bits_per_word = st->bits_per_word;
+	st->offload_xfer.len = (st->bits_per_word > 16) ? 4 : 2;
```
Will check this out for the next version

machschmitt commented:
Wanted some feedback from test results but am updating this anyway since these changes should not break anything.

v3 -> v4:
- No longer register spi_unoptimize_message() devm action.
- Added spi_unoptimize_message() call to ad4630_buffer_postdisable() so  optimize and unoptimize calls are balanced.
- Made SPI message optimize/unoptimize calls while SPI bus is locked to avoid concurrent buffer enable calls possibly calling optimize message on an already optimized SPI message. Not sure if we can have concurrent calls to buffer_preenable() but ...
- Removed experimental patch about lane max sample rate (shouldn't have added that).

dlech commented on drivers/iio/adc/ad4630.c:
```diff
@@ -720,11 +720,17 @@ static int ad4630_buffer_preenable(struct iio_dev *indio_dev)
 	spi_bus_lock(st->spi->master);
-	spi_engine_offload_load_msg(st->spi, &st->offload_msg);
-	spi_engine_offload_enable(st->spi, true);
+	ret = spi_optimize_message(st->spi, &st->offload_msg);
```
FYI, it is OK for `spi_optimize_message()` to be before `spi_bus_lock()`.  It only operates on the message and not the controller.

nunojsa replied:
Yeah, otherwise the lock would also be internal I guess :)

dlech commented:
> * Not sure if we can have concurrent calls to `ad4630_buffer_preenable()` but ...

No, it is not possible. It is already protected by a mutex in the iio core code. (in `iio_update_buffers()`)

nunojsa commented:
> It's not about optimization in this case but the fact that this function compiles the message into SPI Engine instructions.

Ahh, got it. In the offload case we're not going the normal spi_sync() path...

nunojsa commented on drivers/iio/adc/ad4630.c:
```diff
@@ -1197,6 +1203,7 @@ static void ad4630_prepare_spi_sampling_msg(struct ad4630_state *st,
 	}
 
 	st->offload_xfer.bits_per_word = st->bits_per_word;
+	st->offload_xfer.len = (st->bits_per_word > 16) ? 4 : 2;
```
Did you already checked this? I'm fairly sure in the 4 lanes case we can `bits_per_word` as 6 or less (depending on the device)...

machschmitt replied:
If understand it correctly, the number of spi transfer bits should be the current number of precision bits divided by the number of lanes per channel. Except when its more than 16 so 32-bit transfers are made. If so, I think the following should work ...

```c
 #include <linux/bitfield.h>
+#include <linux/bits.h>
 #include <linux/clk.h>
...
-       st->offload_xfer.len = (st->bits_per_word > 16) ? 4 : 2;
+       st->offload_xfer.len = 4;
+       if (st->bits_per_word <= 16)
+               st->offload_xfer.len = DIV_ROUND_UP(st->bits_per_word, BITS_PER_BYTE);
```

@ladace has reported more bugs and there would be more changes to this PR.
I'll wait for this and other changes to be tested before updating this PR again.

dlech replied:
oneliner:

```c
st->offload_xfer.len = roundup_pow_of_two(BITS_TO_BYTES(st->bits_per_word));
```

nunojsa replied:
>  has reported more bugs and there would be more changes to this PR.
I'll wait for this and other changes to be tested before updating this PR again.

Please no... Let's take care of additional bugs in a different PR. No need to add more than needed to this one

machschmitt commented:
Ok, so this PR is now 4 patches only.
- Patch 1 updates adaq4224-24 dts to board revision 3.
- Patch 2 removes average mode zero according to bug report.
- Patch 3 updates ad4630 and adaq4224 dts to new spi-engine implementation
- Patch 4 updates ad4630 driver to the new spi-engine implementation

Change log v4 -> v5
- Brought optimize/unoptimize calls out of SPI bus lock context.
- Took suggested one line expression to assign xfer length.
- Reordered commits to keep spi-engine related changes close one to another.

nunojsa reviewed (approved):
Code LGTM... 

Was it already tested? If not, let me know when we can merge this.

ladace reviewed (approved):
Tested in HW and seems to work, please merge to main!
